src/core/infer_stats.h

```diff
@@ -0,0 +1,132 @@
+        : failure_count_(0), failure_duration_ns_(0), success_count_(0),
+          request_duration_ns_(0), queue_duration_ns_(0),
+          compute_input_duration_ns_(0), compute_infer_duration_ns_(0),
+          compute_output_duration_ns_(0)
+    {
+    }
+    uint64_t failure_count_;
+    uint64_t failure_duration_ns_;
+
+    uint64_t success_count_;
+    uint64_t request_duration_ns_;
+    uint64_t queue_duration_ns_;
+    uint64_t compute_input_duration_ns_;
+    uint64_t compute_infer_duration_ns_;
+    uint64_t compute_output_duration_ns_;
+  };
+
+  struct InferBatchStats {
+    InferBatchStats()
+        : count_(0), compute_input_duration_ns_(0),
+          compute_infer_duration_ns_(0), compute_output_duration_ns_(0)
+    {
+    }
+    uint64_t count_;
+    uint64_t compute_input_duration_ns_;
+    uint64_t compute_infer_duration_ns_;
+    uint64_t compute_output_duration_ns_;
+  };
+
+  // Create an aggregator for model statistics
+  StatsAggregator() : last_inference_ms_(0) {}
+
+  // Set metric reporter for the model statistics
+  void SetMetricReporter(
+      const std::shared_ptr<MetricModelReporter>& metric_reporter)
+  {
+    metric_reporter_ = metric_reporter;
+  }
+
+  uint64_t LastInferenceMs() const { return last_inference_ms_; }
+  const InferStats& ImmutableInferStats() const { return infer_stats_; }
+  const std::map<size_t, InferBatchStats>& ImmutableInferBatchStats() const
+  {
+    return batch_stats_;
+  }
+
+  // Add durations to Infer stats for a failed inference request.
+  void UpdateFailure(uint64_t request_start_ns, uint64_t request_end_ns);
+
+  // Add durations to infer stats for a successful inference request.
+  void UpdateSuccess(
+      uint64_t request_start_ns, uint64_t queue_start_ns,
+      uint64_t compute_start_ns, uint64_t compute_input_end_ns,
+      uint64_t compute_output_start_ns, uint64_t compute_end_ns,
+      uint64_t request_end_ns);
+
+  // Add durations to batch infer stats for a batch execution.
+  // 'success_request_count' is the number of sucess requests in the batch that
+  // have infer_stats attached.
+  void UpdateInferBatchStats(
+      size_t batch_size, uint64_t compute_start_ns,
+      uint64_t compute_input_end_ns, uint64_t compute_output_start_ns,
+      uint64_t compute_end_ns);
+
+ private:
+  std::mutex mu_;
+  uint64_t last_inference_ms_;
+  InferStats infer_stats_;
+  std::map<size_t, InferBatchStats> batch_stats_;
+  std::shared_ptr<MetricModelReporter> metric_reporter_;
+};
+
+//
+// Macros to set infer stats
+//
+#ifdef TRTIS_ENABLE_STATS
+#define INFER_STATS_SET_TIMESTAMP(TS_NS) \
+  {                                      \
+    struct timespec ts;                  \
+    clock_gettime(CLOCK_MONOTONIC, &ts); \
+    TS_NS = TIMESPEC_TO_NANOS(ts);       \
+  }
+#define INFER_STATS_DECL_TIMESTAMP(TS_NS) \
+  uint64_t TS_NS;                         \
+  INFER_STATS_SET_TIMESTAMP(TS_NS);
+#else
+#define INFER_STATS_DECL_TIMESTAMP(TS_NS) uint64_t TS_NS = 0;
```

Comment:
I'm surprised you don't get unused variable errors for this. Can we just leave the #define empty?